In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_excel("Online Retail.xlsx")
print(f"Shape: {df.shape}")
print(f"\nColumn types:\n{df.dtypes}")
print(f"\nMissing values:\n{df.isnull().sum()}")
print(f"\nFirst rows:\n{df.head()}")

Shape: (541909, 8)

Column types:
InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
CustomerID            float64
Country                   str
dtype: object

Missing values:
InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

First rows:
  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country

## Task 1: Data Profiling and Missing Values

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 40.0+ MB


In [4]:
df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [5]:
(df.isna().sum() / len(df)) * 100

InvoiceNo       0.000000
StockCode       0.000000
Description     0.268311
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     24.926694
Country         0.000000
dtype: float64

In [6]:
df.nunique()

InvoiceNo      25900
StockCode       4070
Description     4223
Quantity         722
InvoiceDate    23260
UnitPrice       1630
CustomerID      4372
Country           38
dtype: int64

In [7]:
df.describe()
# quantity and unit price cannot be negative

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,541909.000000,541909,541909.000000,406829.000000
mean,9.552250,2011-07-04 13:34:57.156386,4.611114,15287.690570
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.081158,NaN,96.759853,1713.600303


In [8]:
df[df['Description'].isna()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,2011-12-07 18:26:00,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,2011-12-07 18:31:00,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,2011-12-07 18:35:00,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom


In [9]:
df[df['Description'].notna()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France


In [10]:
df[df['Description'].isna()]['UnitPrice'].sum()

np.float64(0.0)

In [11]:
(df[df['Description'].notna()]['UnitPrice'] == 0).sum()

np.int64(1061)

In [12]:
# Where the description is nan unit price of these are 0

In [13]:
df[df['CustomerID'].isna()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
541536,581498,85099B,JUMBO BAG RED RETROSPOT,5,2011-12-09 10:26:00,4.13,NaN,United Kingdom
541537,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,2011-12-09 10:26:00,4.13,NaN,United Kingdom
541538,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,2011-12-09 10:26:00,4.96,NaN,United Kingdom
541539,581498,85174,S/4 CACTI CANDLES,1,2011-12-09 10:26:00,10.79,NaN,United Kingdom


In [14]:
(df['Country'].value_counts() / len(df)) * 100 # 91% of the data have country UK

Country
United Kingdom          91.431956
Germany                  1.752139
France                   1.579047
EIRE                     1.512431
Spain                    0.467422
Netherlands              0.437527
Belgium                  0.381798
Switzerland              0.369435
Portugal                 0.280305
Australia                0.232327
Norway                   0.200403
Italy                    0.148180
Channel Islands          0.139876
Finland                  0.128250
Cyprus                   0.114779
Sweden                   0.085254
Unspecified              0.082302
Austria                  0.073998
Denmark                  0.071783
Japan                    0.066063
Poland                   0.062926
Israel                   0.054806
USA                      0.053699
Hong Kong                0.053145
Singapore                0.042258
Iceland                  0.033585
Canada                   0.027864
Greece                   0.026942
Malta                    0.023436
United

In [15]:
# 25% of the nan customerid df have country UK
(df[df['CustomerID'].isna()]['Country'].value_counts() / len(df['CustomerID'].isna())) * 100

Country
United Kingdom    24.653586
EIRE               0.131203
Hong Kong          0.053145
Unspecified        0.037276
Switzerland        0.023067
France             0.012179
Israel             0.008673
Portugal           0.007197
Bahrain            0.000369
Name: count, dtype: float64

In [16]:
(df[df['CustomerID'].notna()]['Country'].value_counts() / len(df['CustomerID'].notna())) * 100

Country
United Kingdom          66.778371
Germany                  1.752139
France                   1.566868
EIRE                     1.381228
Spain                    0.467422
Netherlands              0.437527
Belgium                  0.381798
Switzerland              0.346368
Portugal                 0.273109
Australia                0.232327
Norway                   0.200403
Italy                    0.148180
Channel Islands          0.139876
Finland                  0.128250
Cyprus                   0.114779
Sweden                   0.085254
Austria                  0.073998
Denmark                  0.071783
Japan                    0.066063
Poland                   0.062926
USA                      0.053699
Israel                   0.046133
Unspecified              0.045026
Singapore                0.042258
Iceland                  0.033585
Canada                   0.027864
Greece                   0.026942
Malta                    0.023436
United Arab Emirates     0.012548
Europe

In [17]:
df["missing_customer"] = df["CustomerID"].isna().astype(int)
group_stats = df.groupby("missing_customer").agg({
    "Quantity": ["mean","median","std"],
    "UnitPrice": ["mean","median","std"],
})
group_stats

Quantity                    UnitPrice                   
                       mean median         std      mean median         std
missing_customer                                                           
0                 12.061303    5.0  248.693370  3.460471   1.95   69.315162
1                  1.995573    1.0   66.696153  8.076577   3.29  151.900816

In [18]:
df.groupby("missing_customer").agg({
    "Quantity": "mean",
    "UnitPrice": "mean"
})

,Quantity,UnitPrice
missing_customer,,
0,12.061303,3.460471
1,1.995573,8.076577


In [19]:
df["luxury_purchase"] = (
    (df["UnitPrice"] > df["UnitPrice"].quantile(0.9)) &
    (df["Quantity"] <= 2)
).astype(int)

In [20]:
df.groupby("missing_customer")["luxury_purchase"].mean()

missing_customer
0    0.053671
1    0.130841
Name: luxury_purchase, dtype: float64

### The type of the missing values of CustomerID is MAR
Rows with missing CustomerID differ systematically from rows where CustomerID is present. Transactions without CustomerID have significantly lower average quantities (1.99 vs 12.06), higher average unit prices (8.07 vs 3.46), and a higher proportion of luxury purchases (13.08% vs 5.37%). Since the probability of missing CustomerID appears to depend on other observed variables such as price and quantity, the missingness is not completely random. Therefore, the missingness mechanism is best classified as MAR.

### The type of the missing values of Description is MAR
The missingness of the Description field is classified as MAR. My analysis shows that the probability of a description being missing is perfectly predictable based on the UnitPrice variable (where $Price = 0$). These records represent administrative adjustments, damaged stock entries, or internal logs rather than consumer purchases.

## Strategy: Deletion
### Justification:
While deleting these rows removes roughly 24.6% of the dataset, it is the most conservative approach for a customer-centric analysis. Without a valid ID, we cannot perform feature engineering at the customer level.

In [21]:
df.groupby('StockCode')['Description'].nunique().sort_values(ascending=False)

StockCode
20713        8
23084        7
21830        6
85175        6
72807A       5
            ..
DCGS0055     0
DCGS0057     0
DCGS0066P    0
DCGS0071     0
DCGS0072     0
Name: Description, Length: 4070, dtype: int64

In [22]:
df[df['StockCode'] == 20713]['Description'].unique()

array(['JUMBO BAG OWLS', nan, 'wrongly marked. 23343 in box',
       'wrongly coded-23343', 'found', 'Found', 'wrongly marked 23343',
       'Marked as 23343', 'wrongly coded 23343'], dtype=object)

In [23]:
missing_desc_mask = df['Description'].isnull()
missing_desc_df = df[missing_desc_mask]

In [24]:
valid_stock_codes = df.dropna(subset=['Description'])['StockCode'].unique()
missing_with_known_code = missing_desc_df[missing_desc_df['StockCode'].isin(valid_stock_codes)]
missing_with_known_code

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,missing_customer,luxury_purchase
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom,1,0
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom,1,0
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom,1,0
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,NaN,United Kingdom,1,0
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,NaN,United Kingdom,1,0
...,...,...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,2011-12-07 18:26:00,0.0,NaN,United Kingdom,1,0
535326,581203,23406,NaN,15,2011-12-07 18:31:00,0.0,NaN,United Kingdom,1,0
535332,581209,21620,NaN,6,2011-12-07 18:35:00,0.0,NaN,United Kingdom,1,0
536981,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom,1,0


Transactions missing a CustomerID differ systematically from those where the identifier is present, indicating that the data is not missing completely at random. Specifically, transactions without IDs are characterized by significantly lower average quantities—approximately 2 units versus 12 units—and higher average unit prices of roughly 8.07 GBP compared to 3.46 GBP. Geographically, these missing values are heavily concentrated in the United Kingdom, whereas countries like Germany show a 0% missingness rate for IDs. These transactions also show a higher proportion of luxury goods, suggesting they represent retail "guest checkouts" rather than the wholesale-style behavior seen in the registered customer group.

Regarding the second query, transactions with missing descriptions do technically have valid StockCode values, but these codes are often associated with non-standard internal system logs. While it is theoretically possible to recover descriptions by mapping these codes to other rows, it is practically unnecessary because every record with a missing description also has a UnitPrice of 0.0. This correlation confirms that these entries represent administrative errors or stock adjustments rather than commercial sales, making recovery irrelevant for financial analysis. Furthermore, attempting to impute these names would provide little benefit since the lack of pricing data renders these rows unusable for revenue-based modeling or predictive tasks.

In [25]:
df_clean = df.dropna(subset=['Description', 'CustomerID']).copy()

In [26]:
df_clean['CustomerID'] = df_clean['CustomerID'].astype(int)

In [27]:
print(f"Original shape: {df.shape}")
print(f"Cleaned shape:  {df_clean.shape}")
print("\nMissing value counts after cleaning:")
print(df_clean.isnull().sum())

Original shape: (541909, 10)
Cleaned shape:  (406829, 10)

Missing value counts after cleaning:
InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID          0
Country             0
missing_customer    0
luxury_purchase     0
dtype: int64


For Task 1, I applied listwise deletion to all records missing Description or CustomerID values. The missingness in the Description field was classified as MAR because it was perfectly correlated with a UnitPrice of 0.0, indicating that these rows represent non-commercial system adjustments rather than valid sales. Similarly, CustomerID was classified as MAR due to systematic differences in the "Missing" group—specifically lower average quantities, higher unit prices, and a heavy concentration in the UK market compared to the "Known" group. While deleting these records reduces the total dataset size by approximately 25%, it is a necessary step because the absence of a unique customer identifier makes it impossible to perform the longitudinal, customer-level modeling required for the subsequent parts of this lab.

## Task 2: Cleaning Invalid and Anomalous Records

### Decision: Remove them from the primary sales dataset.

Justification: Remove cancellations for the purpose of purchase modeling. Including returns as 'transactions' would create noise in churn prediction. A return is a post-purchase event, not an original purchase intent.

In [28]:
df_clean['IsCancellation'] = df_clean['InvoiceNo'].astype(str).str.startswith('C')

cancel_count = df_clean['IsCancellation'].sum()
print(f"Number of cancellation transactions: {cancel_count}")

Number of cancellation transactions: 8905


In [29]:
df_sales = df_clean[~df_clean['IsCancellation']].copy()

In [30]:
neg_qty_non_c = df_sales[df_sales['Quantity'] <= 0]
zero_price = df_sales[df_sales['UnitPrice'] <= 0]

qty_threshold = df_sales['Quantity'].quantile(0.999)
price_threshold = df_sales['UnitPrice'].quantile(0.999)

extreme_qty = df_sales[df_sales['Quantity'] > qty_threshold]
extreme_price = df_sales[df_sales['UnitPrice'] > price_threshold]

print(f"Quantity <= 0 (Non-Cancellations): {len(neg_qty_non_c)}")
print(f"UnitPrice <= 0: {len(zero_price)}")
print(f"Extreme Quantity Outliers (> {qty_threshold}): {len(extreme_qty)}")
print(f"Extreme Price Outliers (> {price_threshold}): {len(extreme_price)}")

Quantity <= 0 (Non-Cancellations): 0
UnitPrice <= 0: 40
Extreme Quantity Outliers (> 504.0): 393
Extreme Price Outliers (> 41.84239999998827): 398


In [31]:
df_sales = df_sales[(df_sales['Quantity'] > 0) & (df_sales['UnitPrice'] > 0)]

df_sales['is_qty_outlier'] = (df_sales['Quantity'] > qty_threshold).astype(int)
df_sales['is_price_outlier'] = (df_sales['UnitPrice'] > price_threshold).astype(int)

df_sales['Quantity'] = df_sales['Quantity'].clip(upper=qty_threshold)
df_sales['UnitPrice'] = df_sales['UnitPrice'].clip(upper=price_threshold)

print(f"Total outliers flagged: {df_sales['is_qty_outlier'].sum()}")

print(f"New Min Quantity: {df_sales['Quantity'].min()}")
print(f"New Max Quantity: {df_sales['Quantity'].max()}")

Total outliers flagged: 392
New Min Quantity: 1
New Max Quantity: 504


In [32]:
df_sales.shape

(397884, 13)

### Cleaning Summary (Task 2)
Data Removal Decisions

Cancellations: Removed 8,905 transactions (those starting with 'C'). Justification: To focus on successful sales and prevent negative bias in revenue and quantity aggregations.

Invalid Quantities: Removed remaining records with Quantity <= 0. Justification: These represent system errors or non-commercial adjustments that do not reflect actual consumer behavior.

Invalid Prices: Removed records with UnitPrice <= 0. Justification: Free items or price adjustments are excluded to maintain the integrity of revenue-based metrics and high-value customer prediction.

Outlier Management

Strategy: Hybrid Clipping & Flagging. Justification: Instead of removing extreme values (e.g., bulk orders or high-end items), they are flagged using the 99.9th percentile to preserve the signal of "Bulk Buyer" segments. Simultaneously, these values are clipped (Winsorized) to prevent extreme outliers from skewing the mean and destabilizing the revenue-based target engineering in Task 4.

Final Validation

Shape Before: (406,829, 11)

Shape After: (397,884, 13)

Status: Dataset is now clean of invalid transactions and negative values, with extreme values stabilized for predictive modeling.

## Task 3: Categorical Data Challenges

In [33]:
df_sales['Country'].nunique()

37

In [34]:
df_sales['Country'].value_counts(normalize=True).head(5) * 100

Country
United Kingdom    89.051332
Germany            2.272019
France             2.096340
EIRE               1.818621
Spain              0.624303
Name: proportion, dtype: float64

In [35]:
country_counts = df_sales['Country'].value_counts()
rare_countries = country_counts[country_counts < 50]

In [36]:
df_sales['Country_Cleaned'] = df_sales['Country'].apply(lambda x: x if country_counts[x] >= 50 else 'Other')

In [37]:
df_sales['StockCode'].nunique()

3665

In [38]:
df_sales['StockCode']

0         85123A
1          71053
2         84406B
3         84029G
4         84029E
           ...  
541904     22613
541905     22899
541906     23254
541907     23255
541908     22138
Name: StockCode, Length: 397884, dtype: object

In [39]:
non_product_codes = df_sales[df_sales['StockCode'].astype(str).str.contains('^[a-zA-Z]+', regex=True)]['StockCode'].unique()
non_product_codes

array(['POST', 'C2', 'M', 'BANK CHARGES', 'PADS', 'DOT'], dtype=object)

Strategy: Removal from Dataset

For the purpose of product-level analysis, I have chosen to remove all non-product codes, such as POST (postage), DOT (DOTCOM postage), M (manual adjustments), and BANK CHARGES. These entries represent service fees and administrative costs rather than physical retail goods. Including them would distort inventory trends and skew the UnitPrice distribution, potentially misidentifying high-value customers based on their shipping costs rather than their purchasing preferences. Since the downstream goal involves modeling customer behavior, removing these records ensures that our features capture pure product-based intent, eliminating "noise" from non-commercial service adjustments.

## Task 4: Class Imbalance — Predicting High-Value Customers

In [40]:
df_sales['Revenue'] = df_sales['Quantity'] * df_sales['UnitPrice']

customer_df = df_sales.groupby('CustomerID').agg({
    'Revenue': 'sum',
    'InvoiceNo': 'nunique',
    'StockCode': 'nunique',
    'InvoiceDate': ['min', 'max']
}).reset_index()

customer_df.columns = ['CustomerID', 'Total_Revenue', 'Order_Count', 'Unique_Products', 'First_Purchase', 'Last_Purchase']

revenue_threshold = customer_df['Total_Revenue'].quantile(0.90)
customer_df['Is_High_Value'] = (customer_df['Total_Revenue'] >= revenue_threshold).astype(int)

print(f"Revenue Threshold for Top 10%: {revenue_threshold:.2f}")
print(customer_df.head())

Revenue Threshold for Top 10%: 3609.82
   CustomerID  Total_Revenue  Order_Count  Unique_Products  \
0       12346       524.1600            1                1   
1       12347      4310.0000            7              103   
2       12348      1797.2400            4               22   
3       12349      1499.3924            1               73   
4       12350       334.4000            1               17   

       First_Purchase       Last_Purchase  Is_High_Value  
0 2011-01-18 10:01:00 2011-01-18 10:01:00              0  
1 2010-12-07 14:57:00 2011-12-07 15:52:00              1  
2 2010-12-16 19:09:00 2011-09-25 13:13:00              0  
3 2011-11-21 09:51:00 2011-11-21 09:51:00              0  
4 2011-02-02 16:01:00 2011-02-02 16:01:00              0  


In [41]:
distribution = customer_df['Is_High_Value'].value_counts(normalize=True) * 100
print(f"Class Distribution:\n{distribution}")

baseline_accuracy = distribution[0]
print(f"Baseline Accuracy (predicting only 'Regular'): {baseline_accuracy:.2f}%")

Class Distribution:
Is_High_Value
0    89.99539
1    10.00461
Name: proportion, dtype: float64
Baseline Accuracy (predicting only 'Regular'): 90.00%


## Why is accuracy a poor metric here?
Accuracy is misleading because if 90% of customers are "Regular," a model can achieve 90% accuracy by simply guessing "Regular" every time. However, this model would have 0% recall—it would fail to identify a single high-value customer, which is the entire point of the business task. In imbalanced datasets, we prioritize Precision, Recall, and the F1-Score.

In [42]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.utils import resample


X = customer_df[['Order_Count', 'Unique_Products']]
y = customer_df['Is_High_Value']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

train_data = pd.concat([X_train, y_train], axis=1)
regular = train_data[train_data['Is_High_Value'] == 0]
high_value = train_data[train_data['Is_High_Value'] == 1]

high_value_upsampled = resample(high_value, replace=True, n_samples=len(regular), random_state=42)
oversampled_df = pd.concat([regular, high_value_upsampled])

regular_downsampled = resample(regular, replace=False, n_samples=len(high_value), random_state=42)
undersampled_df = pd.concat([regular_downsampled, high_value])

def train_and_eval(train_df, title):
    model = LogisticRegression()
    model.fit(train_df.drop('Is_High_Value', axis=1), train_df['Is_High_Value'])
    preds = model.predict(X_test)
    print(f"\n--- {title} ---")
    print(classification_report(y_test, preds))

train_and_eval(train_data, "Original (No Resampling)")
train_and_eval(oversampled_df, "Oversampled Training Set")
train_and_eval(undersampled_df, "Undersampled Training Set")


--- Original (No Resampling) ---
              precision    recall  f1-score   support

           0       0.95      0.99      0.97       781
           1       0.82      0.52      0.63        87

    accuracy                           0.94       868
   macro avg       0.88      0.75      0.80       868
weighted avg       0.94      0.94      0.93       868


--- Oversampled Training Set ---
              precision    recall  f1-score   support

           0       0.98      0.92      0.95       781
           1       0.54      0.83      0.65        87

    accuracy                           0.91       868
   macro avg       0.76      0.87      0.80       868
weighted avg       0.94      0.91      0.92       868


--- Undersampled Training Set ---
              precision    recall  f1-score   support

           0       0.98      0.92      0.95       781
           1       0.53      0.84      0.65        87

    accuracy                           0.91       868
   macro avg       0.75  

## Interpretation
The results demonstrate a clear trade-off between precision and recall when addressing class imbalance. While the original model achieved the highest accuracy, it failed to identify nearly half of the high-value customers (52% recall). Both oversampling and undersampling significantly improved the model's ability to detect high-value targets, increasing recall to over 80%. Ultimately, Oversampling provided the best balance, as it achieved a higher F1-score than the baseline while utilizing the full extent of the majority class data to maintain a slightly more stable precision than undersampling.

## Task 5: Data Leakage — Introduce, Detect, and Fix

In [43]:
dec_start = pd.Timestamp("2011-12-01")
dec_customers = df_sales[df_sales['InvoiceDate'] >= dec_start]['CustomerID'].unique()
customer_df['Bought_In_Dec'] = customer_df['CustomerID'].isin(dec_customers).astype(int)

X_leak = customer_df[['Total_Revenue', 'Order_Count', 'Unique_Products']]
y_leak = customer_df['Bought_In_Dec']

X_train_l, X_test_l, y_train_l, y_test_l = train_test_split(X_leak, y_leak, test_size=0.2, random_state=42)

model_leak = LogisticRegression()
model_leak.fit(X_train_l, y_train_l)
print("--- Leaked Model Performance ---")
print(classification_report(y_test_l, model_leak.predict(X_test_l)))

--- Leaked Model Performance ---
              precision    recall  f1-score   support

           0       0.89      0.98      0.93       744
           1       0.67      0.25      0.36       124

    accuracy                           0.88       868
   macro avg       0.78      0.61      0.65       868
weighted avg       0.86      0.88      0.85       868



## Detect the Leakage
We get high F1 score and accuracy.

The Leak: The Total_Revenue and Order_Count features include the transactions from December itself.

The "Cheat": If a customer has a high Order_Count calculated from the full year, the model "knows" they were active in December. In a real-world setting, we wouldn't have the "Future Total Revenue" available at the moment you are making a prediction.

In [44]:
obs_end = pd.Timestamp("2011-08-31")
pred_start = pd.Timestamp("2011-09-01")

df_obs = df_sales[df_sales['InvoiceDate'] <= obs_end]
X_clean = df_obs.groupby('CustomerID').agg({
    'Revenue': 'sum',
    'InvoiceNo': 'nunique',
    'StockCode': 'nunique'
}).reset_index()

df_pred = df_sales[df_sales['InvoiceDate'] >= pred_start]
future_customers = df_pred['CustomerID'].unique()
X_clean['Target_Future_Purchase'] = X_clean['CustomerID'].isin(future_customers).astype(int)

X_final = X_clean[['Revenue', 'InvoiceNo', 'StockCode']]
y_final = X_clean['Target_Future_Purchase']

X_train_c, X_test_c, y_train_c, y_test_c = train_test_split(X_final, y_final, test_size=0.2, random_state=42)

model_correct = LogisticRegression()
model_correct.fit(X_train_c, y_train_c)
print("--- Correct Temporal Model Performance ---")
print(classification_report(y_test_c, model_correct.predict(X_test_c)))

--- Correct Temporal Model Performance ---
              precision    recall  f1-score   support

           0       0.61      0.67      0.64       285
           1       0.73      0.69      0.71       378

    accuracy                           0.68       663
   macro avg       0.67      0.68      0.67       663
weighted avg       0.68      0.68      0.68       663



| Split Method | Accuracy | Precision (Class 1) | Recall (Class 1) | F1-Score (Class 1) |
| :--- | :--- | :--- | :--- | :--- |
| **Random (Leaked)** | 0.88 | 0.67 | 0.25 | 0.36 |
| **Temporal (Correct)** | **0.68** | **0.73** | **0.69** | **0.71** |

## Identification of Leaked Information
The leaked model "cheated" by using features like Total_Revenue and Order_Count that were calculated using the entire 12-month dataset. This is a classic case of temporal leakage, where information from the future is used to predict the past. Specifically, if a customer made a purchase in December 2011 (our target), that transaction’s value was already bundled into the "Total Revenue" feature used for training. The model wasn't learning behavioral patterns; it was simply detecting the "echo" of the target event within the features.

## Why the Temporal Split is Correct
The temporal split is the only valid approach for time-series forecasting because it respects the chronological arrow of time. In a real-world business environment, a model can only use data available at the moment of prediction.

By using an "Observation Window" (Dec 2010 – Aug 2011) to create features and a separate "Prediction Window" (Sept 2011 – Dec 2011) for the target, we mimic real-life deployment. Even though the raw accuracy might look lower compared to the leaked model, the F1-score of 0.71 in the correct model represents a far more robust and honest ability to generalize. It shows that the model is actually learning how past frequency and monetary value translate into future loyalty, rather than just spotting future transactions already hidden in the training data.